Build_Process/BootProtocols.md

```diff
@@ -0,0 +1,349 @@
+- `tags`: A pointer to the first stivale2 tag in the linked list of requests.
+
+In the example above we actually set the first tag to a framebuffer request, so lets see what that would look like:
+
+```c
+static stivale2_header_tag_framebuffer framebuffer_tag = 
+{
+    .tag = 
+    {
+        .identifier = STIVALE2_HEADER_TAG_FRAMEBUFFER,
+        .next = 0,
+    },
+    .framebuffer_width = 0,
+    .framebuffer_height = 0,
+    .framebuffer_bpp = 0
+};
+```
+
+The `framebuffer_*` fields can be used to ask for a specific kind of framebuffer, but leaving them to zero tells the bootloader we want to best possible available. The `next` field can be used to point to the next header tag, if we had another one we wanted. The full list of tags is available in the stivale2 specification (see the useful links section).
+
+The last detail is to change the signature of our kernel entry function to:
+
+```c
+void kernel_start(stivale2_struct* stivale2_data);
+```
+
+This struct points to a list of tags, each containing details about the machine we're booted on. These are called struct tags (bootloader -> kernel) as opposed to the tags we defined before (header tags: kernel -> bootloader). To get info about a specific feature, simply walk the linked list of tags, the next tag's address is available in the `tag->next` field. The end of the list is indicated by a nullptr.
+
+## Finding Bootloader Tags
+Since both multiboot 2 and stivale 2 return their info in linked lists, a brief example of how to traverse these lists is given below. These functions provide a nice abstraction to search the list for a specific tag, rather than manually searching each time.
+
+### Multiboot 2
+Multiboot 2 gives us a pointer to the multiboot info struct, which contains 2x 32-bit fields. These can be safely ignored, as the list is null-terminated (a tag with a type 0, and size of 8). The first tag is at 8 bytes after the start of the mbi. All the structures and defines used here are available in the header provided by the multiboot specification (check the bottom section, in the example kernel).
+
+```c
+//placed in ebx when the kernel is booted
+multiboot_info* mbi;
+
+void* multboot2_find_tag(uint32_t type)
+{
+    multiboot_tag* tag = (multiboot_tag*)(uintptr_t)mbi + 8);
+    while (1)
+    {
+        if (tag->type == 0 && size == 8)
+            return NULL; //we've reached the terminating tag
+        
+        if (tag->type == type)
+            return tag;
+        
+        tag = (multiboot_tag*)((uintptr_t)tag + tag->size);
+    }
+}
+```
+
+The last line of the loop is a little messy, in fact most compilers will let you omit the final cast to the `void*`, and will convert the integer expression directory to a pointer for you (gcc and clang both allow this).
+
```

Comment:
In the original multiboot file:
**Tags follow one another padded when necessary** in order for each tag to start at 8-bytes aligned address. So when iterating through tags we need always to align the address to 8byte, this explain the for loop in the multiboot2 example:

```
  for (tag = (struct multiboot_tag *) (addr + 8);
       tag->type != MULTIBOOT_TAG_TYPE_END;
       tag = (struct multiboot_tag *) ((multiboot_uint8_t *) tag 
                                       + ((tag->size + 7) & ~7)))
```
And the text says: 
To get to the next tag, we always need to add the size of the current tag (that is in tag->size) and align it to 8bytes. this is done by adding 7 to the current tag size, and then AND the result with negation of 7.

Tags are terminated by a tag of type ‘0’ and size ‘8’.

One thing, i'm totally ok with the while loop, but please mention the aligned  increment part (i took it for myself but it took time for me to figure out what was the meaning of this for loop increment! :D And why that + 7 & ~7

But what i was wondering is that it says that the header are padded when necessary, So in theory the while could not always end at the beginning of the next item if it doesn't take it into account? (maybe you should keep the aligned increment in the while too?)